core/src/main/java/hudson/lifecycle/ExitLifecycle.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.lifecycle;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import jenkins.model.Configuration;
+import jenkins.model.Jenkins;
+
+/**
+ * {@link Lifecycle} implementation when Hudson runs on the embedded
+ * servlet container.
+ *
+ * <p>
+ * Restart by exit with specific code.
+ *
+ * @author Alon Bar-Lev
+ */
+public class ExitLifecycle extends Lifecycle {
+
+    private static final String EXIT_CODE_ON_RESTART = "exitCodeOnRestart";
+    private static final String DEFAULT_EXIT_CODE = "5";
```

Comment:
Is there any use-case for specifying the exit code used? I would argue it should be 0, by default at least.

Reply:
We must provide a way to distinguish between stop and restart. Stop is exit code 0, restart should be something else, I use 5 as default it can be anything, I can put 55 as in the commit message to emphasis it as unique.

Setting it to 0 will result in stop instead of restart, maybe this what you want.